### DISCUSSION TITLE
How to work go forward with the CLI in terms of client paramters (required & optional)

### DISCUSSION THREAD

**Discussion by StefMa** [ioki-mobility/go-outline#58] [General]
Right now, we only support required parameters for some of our http client provided functionality.
This is done via:
`APPNAME COMMAND SUBCOMAND ARGS GLOBALFLAGS`
Example:
`outline document create docName collectionId --server http.. --key apIkey`

My questions are:
1. Is this a good style? 
Meaning: Should we use ARGS for the required parameters? 🤔 
2. How to add optional paraemters
Should we go with `localFlags` (for that specific command) in this case?
3. Should we go with `localFlags` for all params? Required as well as optional? Should we simly omit args? 🤔 
If so: For what should we use args? 🤔 

I hope my questions are clear enough.
If not, I'm happy to give more details about my thinking. Maybe even provide examples...

**Comment by rsjethani** [comment]
Hey thanks for raising this discussion, at some point this needs to be addressed.

The optional args/flags like `--server` at the moment look counter intuitive because we are asking something compulsory but via optional flag format.

If we implement proper 12 factor app then server and api key can come from following sources with increasing priority:
1. default hard coded values
2. optional config file say a JSON/YAML config file
3. optional env vars say OCLI_SERVER  and OCLI_API_KEY
4. finally via command line args <-- this is where we are

So  `--server` and `--key` and are actually meant to **override** these mandatory values which should primarily come from env var or config file in the first place. For example aws cli and github cli maintain user specific config file in user's home directory.

**Reply by StefMa** [reply]
Thanks. But this doesn't answer my initial question 😁 I especially talk about optional parameters in api calls. For example the "publish" parameter in the "create document" call. It can be set to true or false. How to model this in the cli?

**Comment by StefMa** [comment]
I just read this one https://actuated.dev/blog/develop-a-great-go-cli and I have the feeling we should use more flags and literally no args 😁

**Reply by fabrik42** [reply]
I would argue, semantically there is a difference between a API key and server host and an ID.

Following the explanation of @rsjethani above:
* Flags are a fallback for other config options
* For an operation like "Get document" you would never store the (share) id in an environment variable.
Because it is not a config value, but an arg for a single operation. 
* Therefore it would make more sense to pass the (share) id as an arg, not as a flag

**Reply by StefMa** [reply]
Thanks for your opinion about this!
However, I guess it make sense to start with examples.

This is how the [official cobra website](https://cobra.dev/) describes Commands, Args and Flags
> Commands represent actions, Args are things and Flags are modifiers for those actions.

So we can conclute, *flags* modify the *actions*.

Lets take the [**create document**](https://github.com/ioki-mobility/go-outline/pull/59) as an example.
Currently it looks like this:
> ./outline document create documentFromCli 633e497f-8a41-482a-937e-0d00ee326963 --server [URL] --key [KEY]

While:
`document` = command
`create` == (sub)command
`documentFromCli` == arg
`633e497f-8a41-482a-937e-0d00ee326963` == arg
server & key == flag

Right now, the arguments has to be in an fixed order!
The first one is always **the name** of the document, while the second one is always the **collection id** where the document will be created.
It doesn't work the other way around (first = collection id, second = document title).
Code ref: https://github.com/ioki-mobility/go-outline/pull/59/files#diff-223a11f450a34cbbac0fdfc3e1f9caeefd948071a628813149e6f8aa7c2bc15eR102

Right now, we also **don't support** optional parameters (from API perspective).
Like the `publish` boolean parameter. 
Code ref: https://github.com/ioki-mobility/go-outline/blob/e1ebe6bfa314903a010ef6483ee24134aef9497f/documents.go#L99-L102
API ref: https://www.getoutline.com/developers#tag/Documents/paths/~1documents.create/post

My question is: 
"Where" to place these optional (like `publish`) to the CLI? 🤔 

Should it be:
1. An additional arg (that is optional)?
2. An flag (that has even a nice name)?
3. Something else

As we agree that `flags` are things to "modify" the `action` (therefore the `create` command), I guess it should be an arg, right? 🤔 
But if this is the case, how do we parse all of the optional args in a way users can omit them? 🤔 
Even if we stick to the "first two awlays has to be in the same order", then all of the following optionls were be possible for all of our optional API parameters:

> create documentFromCli 633e497f-8a41-482a-937e-0d00ee326963 publish --server [URL] --key [KEY]
> create documentFromCli 633e497f-8a41-482a-937e-0d00ee326963 publish templateId --server [URL] --key [KEY]
> create documentFromCli 633e497f-8a41-482a-937e-0d00ee326963 templateId publish --server [URL] --key [KEY]
> create documentFromCli 633e497f-8a41-482a-937e-0d00ee326963 text templateId publish --server [URL] --key [KEY]
> create documentFromCli 633e497f-8a41-482a-937e-0d00ee326963 publish text templateId --server [URL] --key [KEY]
> ...
> What not / I guess you get the point

My question is:
How should the CLI distringuish between those args? 
How does it know that the args can be placed in any order? 

**Therefore**, I thought to set those optional parameters as `flags`.
So we would end up with:
> create documentFromCli 633e497f-8a41-482a-937e-0d00ee326963 --publish true/false --server [URL] --key [KEY]
> create documentFromCli 633e497f-8a41-482a-937e-0d00ee326963 --publish true/false templateId ID --server [URL] --key [KEY]
> create documentFromCli 633e497f-8a41-482a-937e-0d00ee326963 --templateId ID --publish true/false --server [URL] --key [KEY]
> create documentFromCli 633e497f-8a41-482a-937e-0d00ee326963 --text "ABC" --templateId ID --publish true/false --server [URL] --key [KEY]
> create documentFromCli 633e497f-8a41-482a-937e-0d00ee326963 --publish true/false --text "ABC" --templateId ID --server [URL] --key [KEY]
> ...
> What not / I guess you get the point

However, this would **not match** with the statement:
> Commands represent actions, Args are things and Flags are modifiers for those actions.

As the Flags are not modifier for those actions (command == `create` in this case), but rather "things" (like `args` should be 😁 ).
Right? 🤔

**Reply by fabrik42** [reply]
Hey, thanks for the write-up! I find this discussion quite interesting!

I agree with your reasoning.

However, I also agree that it is also somehow uncool to use flags to pass "all kinds of data" (like text, header, etc.) instead of some simple modifiers.

When it comes to things like markdown content,  a simple CLI might not be the best way to pass this data.

We could also think about other ways, like:

1. you feed it a json file with this config and content
2. you can pipe content and/or config into our CLI
3. you can place this kind of metadata somewhere else e.g. in a markdown file.

What I was hoping for:
If Outline would support [FrontMatter](https://jekyllrb.com/docs/front-matter/) like Jekyll or Hugo does, we could use this to store this kind of information, just like most static site generators do.

So a markdown file could look like this:


```md
---
published: true
templateId: 123
---

# This is the actual content

Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum.
```

But I am not sure if this is planned or wanted.

**Comment by rsjethani** [comment]
My experience as both CLI consumer and developer:

**root command i.e. the app name itself**: responsible for optional flags to override some config behavior that is to be inherited by all other actions/commands (--key and --server in our case). For simpler apps where there is only main command/action to perform, the root command itself can do this. For eg starting a server like app or the cli has only one possible action say copying files (the `cp` command).

**first level commands**: start a new scope which inherit global configuration, may also have actions if there are no sub-commands

**(sub)commands**: are actual actions with optional flags to modify/augment the main action/purpose

**arguments**: are mandatory values that are required to complete the action of a command. They are **positional** in nature hence the **MUST** be given in the defined order. If the required data is too complex or too big to be simply passed as an argument then we could 1/ read from files 2/  read from stdin or 3/ go to interactive mode and get values via QnA or allowing copy paste.

**Reply by rsjethani** [reply]
Based on above lets deal with some outline cli cases:

**Getting a document by uuid or a share id**
Here only one value is required i.e. only one positional argument. We can the argument as uuid value by default and make use of an optional flag to treat differently to modify action behavior. For example we can have `-s`/`--share` optional flash to treat the argument as a share id. So final cli structure may look like: `oulcli document get [-s] UUID_OR_SHAREID`.

**Reply by rsjethani** [reply]
Next case: **Creating a document**
This is a bit complex as we can see from discussions :smile: but lets see. Here we need use all cli techniques at our disposal to build the final request:
- `title` is a required value hence this would be the FIRST positional argument
- similarly `collection id` would be second positional argument
- by default there is no `parent document id` but  can be overridden by `--pdi` flag
- by default there is no `template id`  but can be overridden by `--tid` flag
- by default do not treat as template but can override via `--template` flag
- by default we do not publish but can be overridden by `--publish` flag

How do we ask for document text?
Well we can ask for a file name via a flag say `--text-file` then read from that file. We can go even further and have another _mutually exclusive_ option `--text-stdin` indicating that we are supposed to read text via stdin.

NOTE: all optional flags discussed so far are ultimately modifying/augmenting how the action: "how the document is created" hence in my eyes they are not violating my understanding of `cli flags/options`.

So final cli structure may look like:
`outcli document create [--pdi val] [--tid val] [--publish] [--template] [--text-file val|--text-stdin] TITLE COLLECTION_ID`

Bonus tip: The titles can usually contain spaces  i.e. it is mostly free text so the user is forced to quote the whole arg so that 2nd pos arg can be read correctly.Hence  Making title as second arg is better UX. As a result after giving all optional flags and the first pos arg everything thing left as free text is the tile. For eg: `outcli document create --publish 14234-234-234 my new document`

**Reply by rsjethani** [reply]
Sometimes values may really fall into the grey area where it may not be easy to decide on arg vs flag etc.  This may be due to the design choices of the target system/api because ultimately the CLI is a reflection of the remote System/API it wants to interact with.

**Comment by StefMa** [comment]
Alight alight, lets consider this topic as done for now.
After all your impressions and feedback (thanks for that!) and after a discussion via Slack with @rsjethani we decided to go the following way:

| HTTP params | CLI params |
|-|-|
| required | arguments |
| optional | flags |

So basically what is also written [here](https://github.com/ioki-mobility/go-outline/discussions/58#discussioncomment-7088783).

We also had a look at the `gh` (GitHub) cli  and they also use `flags` as "HTTP parameters".
See for example: 
https://cli.github.com/manual/gh_issue_create
https://cli.github.com/manual/gh_pr_close

However, it is also interactive:
![ezgif com-gif-maker(1)](https://github.com/ioki-mobility/go-outline/assets/10229883/cd6ed33b-8112-4d6a-b064-65c52914ef9f)

We might consider this in the future as well.
Using https://github.com/go-survey/survey or https://github.com/charmbracelet/bubbletea.
